### PR TITLE
Separating Graph methods for adding nodes and compiling graph

### PR DESCRIPTION
**Pull request by baykaner**
- Reworked graph to split AddingNodes and graph compilation. This permits adding nodes in an arbitrary order to the graph.
- Reworked graph Evaluate/ForwardPropagate to be defensive (i.e. ForwardPropagate is now a protected member)

### CODE REVIEW

File: libs/ml/include/ml/core/graph.hpp [hunk 1]
```diff
@@ -97,82 +130,175 @@ class Graph
...
+      std::string subnode_name(node_name + "_" + trainable.first);
+
+      // only add new trainable
+      // graph re-compilation can lead to a valid call to add the same trainables twice, this should
+      // be ignored
+      if (graph_ptr->trainable_lookup_.find(subnode_name) == graph_ptr->trainable_lookup_.end())
+      {
+        //        graph_ptr->trainable_nodes_.emplace_back(graph_ptr->trainable_nodes_.at(trainable.second));
+        //        graph_ptr->trainable_lookup_[subnode_name] = graph_ptr->trainable_nodes_.size() -
+        //        1;
```
Comment (MissingNO57) [inline libs/ml/include/ml/core/graph.hpp]: Commented out leftover code

File: libs/ml/include/ml/core/graph.hpp [hunk 2]
```diff
@@ -195,164 +323,124 @@ TensorType Graph<TensorType>::Evaluate(std::string const &node_name, bool is_tra
...
-void Graph<TensorType>::BackPropagateError(std::string const &node_name)
+TensorType Graph<TensorType>::ForwardPropagate(std::string const &node_name, bool is_training)
 {
-  TensorType error_signal;
+  Compile();
+
   if (nodes_.find(node_name) != nodes_.end())
   {
-    nodes_[node_name]->BackPropagateSignal(error_signal);
+    return ((*(nodes_[node_name]->Evaluate(is_training))));
```
Comment (MissingNO57) [inline libs/ml/include/ml/core/graph.hpp]: Why there is so many brackets?

File: libs/ml/include/ml/core/node.hpp [hunk 3]
```diff
@@ -257,6 +258,17 @@ std::vector<std::pair<Node<T> *, T>> Node<T>::BackPropagateSignal(TensorType con
   return non_back_propagated_error_signals;
 }
 
+/**
+ * Resets input and output node ptr containers. Useful for graph decopmiling.
```
Comment (evsmithx) [inline libs/ml/include/ml/core/node.hpp]: Spelling

File: libs/ml/tests/ml/graph/graph.cpp [hunk 4]
```diff
@@ -100,7 +100,39 @@ TYPED_TEST(GraphTest, no_such_node_test)  // Use the class as a Node
   TensorType data(std::vector<SizeType>({5, 10}));
   g.SetInput("Input", data);
 
-  ASSERT_ANY_THROW(g.ForwardPropagate("FullyConnected"));
+  ASSERT_ANY_THROW(g.Evaluate("FullyConnected"));
+}
+
+TYPED_TEST(GraphTest, node_add_wrong_order_test)  // Use the class as a Node
```
Comment (evsmithx) [inline libs/ml/tests/ml/graph/graph.cpp]: Incorrect comment?

**Review by MissingNO57 — CHANGES_REQUESTED** [review]
Just some minor style issues.

**Review by MissingNO57 — DISMISSED** [review]
LGTM

**Review by MissingNO57 — APPROVED** [review]
LGTM

**Pull request merged**